src/Storages/StorageURL.cpp

```diff
@@ -219,8 +219,33 @@ StorageURLSink::StorageURLSink(
     const String & http_method)
     : SinkToStorage(sample_block)
 {
+    //
+    // get the content type first
+    //
+    // The code here may look a little wired.
+    // The getContentType() is prodived on IOutputFormat class which relies on a WriteBuffer object,
+    // and this WriteBuffer object here is WriterBufferFromHTTP itself which accepts the Content-Type header.
+    // So, this is cyclic dependency. 
+    // To decouple such dependency, we must be able to set header to 'WriteBufferFromHTTP' after we get the instance of output format by calling IOutputFormat::getContentType.
+    // But this is tricky because the 'WriteBufferFromHTTP' object may have been decorated by 'WriteBufferWithCompression' and is not acceesible due to private modifiers.
+    //
+    // So, here we first instantiate an OutputFormat object with a fake stream to get the Content-Type.
+    // This is not the best way but a more simpler way to understand.
+    //
+    std::string content_type;
+    {
+        WriteBufferFromOStream buffer(std::cout);
+        auto output = FormatFactory::instance().getOutputFormat(format, 
```

Comment:
This way is a too huge hack.
If you need content type for format before you actually get output format then it is better to add a method to `FormatFactory.h` like `getFormatContentType(const String & name)`. Should be implemented the same way as there is now `bool checkIfFormatIsColumnOriented(const String & name)`.

Reply:
@kssenii Thanks for your feedback. I know it's not the best way. 

As your suggestion, even if we add a new `getFormatContentType` to `FormatFactory`, but how could we get the content type without instantiating an instance of `IOutputFormat`? 

`checkIfFormatIsColumnOriented` returns the `is_column_oritented` field of `FormatFactory::Creators`, do you mean that we add an extra field to this `Creator` to hold the content type information?

Reply:
> do you mean that we add an extra field to this Creator to hold the content type information?

Yes.

Reply:
There's a problem to do that. Some Content-Type is not a static string but is determined during runtime. For example, 

```
    String getContentType() const override
    {
        return String("text/csv; charset=UTF-8; header=") + (with_names ? "present" : "absent");
    }
```

and my another PR: https://github.com/ClickHouse/ClickHouse/pull/32112

Reply:
Then may be try to put this logic with getting output format and then returning content type as a method in format factory. May be if possible try to collect minimum what is needed for that from getOutputFormat method.

Reply:
If we do this, the content-type have two copies, one is the `getContentType` defined on `IOutputFormat` and its sub-classes, another is the new field on `Creator `.  And of course, we need to eliminate the old one, which will introduce more changes, I don't know if we need to do this in this PR. We usually to such refactoring in a following up PR, what do you think?

Reply:
>  another is the new field on Creator

I had an impression the we dropped this field in creator because there is not point in it as it cannot be statically defined when format is registered.

Reply:
> I don't know if we need to do this in this PR. We usually to such refactoring in a following up PR, what do you think?

Well, may be, but current solution is too tricky, I am very doubtful about merging it.

Reply:
I am suggesting to add a method in format factory `getFormatContentType(name)` which will have inside a minimum version from getOutputFormat - monimum to be able to call getContentType in result format and return it back.

Reply:
I see. I think this function would look like as follows:

```c++
String FormatFactory::getOutputFormat(
    const String & name,
    ContextPtr context,
    const std::optional<FormatSettings> & _format_settings) const
{
    const auto & output_getter = getCreators(name).output_creator;

    auto format_settings = _format_settings ? *_format_settings : getFormatSettings(context);
    WriteBufferFromOStream emptyBuf(std::cout);
    RowOutputFormatParams params;
    Block emptyBlock;
    auto format = output_getter(emptyBuf, emptyBlock, params, format_settings);

    return format->getContentType();
}
```

I have not checked if the `emptyBlock` above works.

Reply:
With block it wont, but the general idea is correct.